### PR TITLE
Compile arrays

### PR DESCRIPTION
**Pull request by ronbel**

### CODE REVIEW

File: src/ex2/CompileVisitor.java [hunk 1]
```diff
@@ -196,19 +197,135 @@ public void visit(WhileStatement whileStatement) {
...
+
+        assignArrayStatement.index().accept(this);
+        var indexRegister = lastRegisterNumber-1;
+        var indexComparisonToZeroRegister = lastRegisterNumber++;
+        addLine(String.format("%%_%s = icmp slt i32 %%_%s, 0", indexComparisonToZeroRegister, indexRegister));
+
+        var label1 = lastLabelNumber++;
+        var label2 = lastLabelNumber++;
+        addLine(String.format("br i1 %%_%s, label %%arr_check%s, label %%arr_check%s", indexComparisonToZeroRegister, label1, label2));
+        addLine(String.format("arr_check%s:", label1));
```
Comment (LiadOz) [inline src/ex2/CompileVisitor.java]: function addLine also puts a tab at the start and it's be easier to read if labels do not have tabs

Comment (ronbel) [reply src/ex2/CompileVisitor.java]: Ok, will fix

File: src/ex2/CompileVisitor.java [hunk 2]
```diff
@@ -196,19 +197,135 @@ public void visit(WhileStatement whileStatement) {
...
+
+        var allocationRegister = lastRegisterNumber++;
+
+        addLine(String.format("%%_%s = call i8* @calloc(i32 4, i32 %%_%s)", allocationRegister, lengthPlusOneResultRegister));
+
+        var castRegister = lastRegisterNumber++;
+
+        addLine(String.format("%%_%s = bitcast i8* %%_%s to i32*", castRegister, allocationRegister));
+
+        addLine(String.format("store i32 %%_%s, i32* %%_%s", lengthResultRegister, allocationRegister));
```
Comment (LiadOz) [inline src/ex2/CompileVisitor.java:355]: last line should have a pointer to the array so it can be accessed after creation

**Comment by LiadOz** [comment]
I think you missed adding the throw_oob template

File: src/ex2/CompileVisitor.java (hunk 2 above)
Comment (ronbel) [reply src/ex2/CompileVisitor.java:355]: The last line doesn't allocate a new register, so the last register (lastRegisterNumber-1) will be the register that holds the array

**Comment by ronbel** [comment]
> I think you missed adding the throw_oob template

We can add it later on because the helper methods will be present in every file anyway. Do you want me to add it in my branch?

**Comment by LiadOz** [comment]
> 
> 
> > I think you missed adding the throw_oob template
> 
> We can add it later on because the helper methods will be present in every file anyway. Do you want me to add it in my branch?

seems to me it's best we add it here

**Comment by ronbel** [comment]
> > > I think you missed adding the throw_oob template
> > 
> > 
> > We can add it later on because the helper methods will be present in every file anyway. Do you want me to add it in my branch?
> 
> seems to me it's best we add it here

Done

**Pull request merged**